spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -93,7 +94,8 @@
 
 	private final RedisTemplate<String, RedisLock> redisTemplate;
 
-	private final ThreadLocal<List<RedisLock>> threadLocks = new ThreadLocal<List<RedisLock>>();
+	private final ThreadLocal<Set<RedisLock>> weakThreadLocks = new ThreadLocal<Set<RedisLock>>();
+	private final ThreadLocal<List<RedisLock>> hardThreadLocks = new ThreadLocal<List<RedisLock>>();
```

Comment:
One big problem: hash code need to be immutable since you put into local hash based storage - but it does not (coz of `lockedAt` and `threadName`). Second, tiny one: this statement sometimes may produce false `registry.obtain('key').equals(registry.obtain('key'))` cos of GC. My first attempt was exactly the same as you did, and I gave up cos of mutability of the hash, so I decided just move it between maps and keep previous approach to have same lock instance while it is locked...